props/websubmit/src/main.rs

```diff
@@ -177,6 +177,112 @@ fn run_sc_policy(ctx: Arc<Context>) -> Result<()> {
     })
 }
 
+/// If sensitive data is released, the release must be scoped, and all inputs to the scope must be safe.
+pub struct AuthDisclosureProp {
+    cx: Arc<PolicyContext>,
+}
+
+impl AuthDisclosureProp {
+    pub fn new(cx: Arc<PolicyContext>) -> Self {
+        AuthDisclosureProp { cx }
+    }
+
+    pub fn check(self) {
+        for c_id in self.cx.desc().controllers.keys() {
+            // All srcs that have no influencers
+            let roots = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|n| {
+                    self.cx
+                        .influencers(*n, paralegal_policy::EdgeType::DataAndControl)
+                        .next()
+                        .is_none()
+                })
+                .collect::<Vec<_>>();
+
+            let safe_scopes = self
+                .cx
+                // All nodes marked "safe"
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|n| self.cx.has_marker(marker!(safe_source), *n))
+                // And all nodes marked "safe_with_bless"
+                .chain(self.cx.all_nodes_for_ctrl(*c_id).filter(|node| {
+                    self.cx.has_marker(marker!(safe_source_with_bless), *node)
+                        && self
+                            .cx
+                            // That are influenced by a node marked "bless"
+                            .influencers(*node, paralegal_policy::EdgeType::DataAndControl)
+                            .any(|b| self.cx.has_marker(marker!(bless_safe_source), b))
+                }))
+                .collect::<Vec<_>>();
+            let sinks = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|n| self.cx.has_marker(marker!(sink), *n))
+                .collect::<Vec<_>>();
+            let mut sensitives = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(sensitive), *node));
+
+            let controller_valid = sensitives.all(|sens| {
+                sinks.iter().all(|sink| {
+                    // sensitive flows to store implies
+                    !(self
+                        .cx
+                        .flows_to(sens, *sink, paralegal_policy::EdgeType::Data))
+                        || sink.associated_call_site().is_some_and(|sink_callsite| {
```

Comment:
One way I would suggest to make easier to read is to not use `||` here but `if flows_to(...) { return ... }`. That way you have less indentation later.

PS: I also really like to write highly nested, complex expression code, but I think we endeavor to make our code legible 🙄 .

Reply:
done